[GeoMechanicsApplication] Clean up retention response

avdg81 opened the pull request:
**📝 Description**
Cleaned up several things related to the calculation of the retention response.

**🆕 Changelog**
- When calculating the values at integration points, only call the relevant functions rather than `CalculateRetentionResponse` (which calculates all retention parameters). This saves us from calculating (and storing) unnecessary data.
- Consequently, it was no longer needed to calculate the effective saturation as part of `CalculateRetentionResponse`.
- Removed members `FluidPressure` and `EffectiveSaturation` from struct `UPwSmallStrainElement::ElementVariables`, since they had become redundant. Their usages have been replaced by adopting local variables. In a similar way, members `FluidPressure` and `Density` have been removed from struct `SmallStrainUPwDiffOrderElement::ElementVariables`.
- Removed two unused `using` statements.

avdg81 committed:
Replaced two calls to `CalculateRetentionResponse` by more specific ones

`CalculateRetentionResponse` calculates several parameters, even when
not all of them are needed. In these two instances, it was clear how to
replace these function calls, which makes the code easier to understand.

avdg81 committed:
Removed two redundant `using` statements

avdg81 committed:
Use a local variable for the Bishop coefficient when possible

This eliminates two usages of member `BishopCoefficient` of struct
`ElementVariables`.

avdg81 committed:
Eliminated member `FluidPressure` of struct `ElementVariables`

Its usages could be replaced by using local variables.

avdg81 committed:
Removed member `EffectiveSaturation` of struct `ElementVariables`

It was assigned to only once, but never read from (due to earlier
changes).

avdg81 committed:
Replaced a call to `CalculateRetentionResponse` by a more specific one

In this case, only the fluid pressure and the Bishop coefficient were
needed.

avdg81 committed:
Removed member `FluidPressure` from struct `ElementVariables`

Replaced all of its usages by local variables. Also replaced two usages
of member `BishopCoefficient` by a local variable.

avdg81 committed:
Removed member `Density` from struct `ElementVariables`

Its single usage has been replaced by a local variable.

avdg81 committed:
Fixed GCC builds by adding `this->` to `CalculateRetentionResponse`

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -520,15 +520,22 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateOnIntegrationPoints(const
             if (rVariable == RELATIVE_PERMEABILITY)
-                rOutput[GPoint] = Variables.RelativePermeability;
+                rOutput[GPoint] =
+                    mRetentionLawVector[GPoint]->CalculateRelativePermeability(RetentionParameters);
```
These are mutually exclusive.

avdg81 replied:
Yes, I agree with your observation. I have changed the other branches to use `else if`. I would, however, argue that it would be better still to have such selections one level higher. It is no longer needed to group these variables, since we no longer call `CalculateRetentionResponse`. But I would suggest to do that another time.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -714,11 +721,14 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateOnIntegrationPoints(const
             noalias(TotalStressVector) = mStressVector[GPoint];
             noalias(TotalStressVector) += PORE_PRESSURE_SIGN_FACTOR * Variables.BiotCoefficient *
-                                          Variables.BishopCoefficient * Variables.FluidPressure * VoigtVector;
+                                          bishop_coefficient * fluid_pressure * VoigtVector;
```
sig_tot = sig_eff + p * I 
is computed here. That is probably clearer written down in 1 statement.

avdg81 replied:
Yes, again I agree with you. I have made this calculation more straightforward and I've tried to take out any unnecessary intermediate variables.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1088,9 +1098,9 @@ std::vector<array_1d<double, TDim>> UPwSmallStrainElement<TDim, TNumNodes>::Calc
+        const auto fluid_pressure =
             GeoTransportEquationUtilities::CalculateFluidPressure(Variables.Np, Variables.PressureVector);
-        RetentionParameters.SetFluidPressure(Variables.FluidPressure);
+        RetentionParameters.SetFluidPressure(fluid_pressure);
```
Why not like RetentionParameter.SetFluidPressure(GeoTransportEquationUtilities::CalculateFluidPressure(Variables.Np, Variables.PressureVector));

avdg81 replied:
I thought it might be somewhat too long of an expression. But clearly, your perception is different, so I'll change it :-)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1628,18 +1637,16 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateRetentionResponse(ElementV
         GeoTransportEquationUtilities::CalculateFluidPressure(rVariables.Np, rVariables.PressureVector);
-    rRetentionParameters.SetFluidPressure(rVariables.FluidPressure);
+    rRetentionParameters.SetFluidPressure(fluid_pressure);
 
```
See previous comment.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:
```diff
@@ -923,9 +923,9 @@ void SmallStrainUPwDiffOrderElement::CalculateOnIntegrationPoints(const Variable
                 Variables.Np, Variables.PressureVector);
-            RetentionParameters.SetFluidPressure(Variables.FluidPressure);
+            RetentionParameters.SetFluidPressure(fluid_pressure);
 
```
Also 3 of these changes here in this file ( see 1033, 1167, 2957). If possible, do it in 1 statement.

avdg81 replied:
Done.

WPK4FEM reviewed (dismissed):
Short and comprehensible. Remarks are improvementpoints not requirements.

avdg81 committed:
Processed Wijtze Pieter's review comments

- Made some selections mutually exclusive.
- Eliminated a few intermediate variables.
- Made the calculation of the total stress vector more straightforward.
- Removed a few unnecessary resize operations.

rfaasse reviewed (dismissed):
This is a very nice and straight-forward PR and a clear improvement, I don't have anything to add 😄

Probably we can slowly but surely stop using the `CalculateRetentionResponse` everywhere, but this is a nice start!

WPK4FEM reviewed (approved):
Fine for me.

rfaasse reviewed (approved):
Ready to go!
